controllers/vaultsecret_controller.go

```diff
@@ -0,0 +1,307 @@
+			if defstat.LeaseDuration < smallestLeaseDurationSeconds {
+				smallestLeaseDurationSeconds = defstat.LeaseDuration
+			}
+		}
+
+		//No lease durations found
+		if smallestLeaseDurationSeconds == math.MaxInt64 {
+			return -1, false
+		}
+
+		percentage := float64(instance.Spec.RefreshThreshold) / float64(100)
+		scaledSeconds := float64(smallestLeaseDurationSeconds) * percentage
+		duration := time.Duration(scaledSeconds) * time.Second
+		return duration, true
+	}
+
+	//No refresh period or definitions status known
+	return -1, false
+
+}
+
+func (r *VaultSecretReconciler) manageReconcileLogic(ctx context.Context, instance *redhatcopv1alpha1.VaultSecret) error {
+
+	duration, ok := r.calculateDuration(instance)
+
+	// if this has reconciled before
+	if instance.Status.LastVaultSecretUpdate != nil {
+		// if the next duration is incalculable (no refreshperiod or lease duration), do not reconcile
+		if !ok {
+			return nil
+		}
+		// if the resync period has not elapsed, do not reconcile
+		if !instance.Status.LastVaultSecretUpdate.Add(duration).Before(time.Now()) {
+			return nil
+		}
+	}
+
+	r.Log.V(1).Info("Reconcile VaultSecret", "namespacedName", fmt.Sprintf("%v/%v", instance.Namespace, instance.Name))
+
+	mergedMap := make(map[string]interface{})
+
+	definitionsStatus := make([]redhatcopv1alpha1.VaultSecretDefinitionStatus, len(instance.Spec.VaultSecretDefinitions))
+
+	for idx, vaultSecretDefinition := range instance.Spec.VaultSecretDefinitions {
+		vaultClient, err := vaultSecretDefinition.Authentication.GetVaultClient(ctx, instance.Namespace)
+		if err != nil {
+			r.Log.Error(err, "unable to create vault client", "instance", instance)
+			return err
+		}
+
+		ctx = context.WithValue(ctx, "vaultClient", vaultClient)
+		vaultEndpoint := vaultutils.NewVaultEndpointObj(&vaultSecretDefinition)
+		vaultSecret, ok, _ := vaultEndpoint.ReadSecret(ctx)
+		if !ok {
+			return errors.New("unable to read Vault Secret for " + vaultSecretDefinition.GetPath())
+		}
+
+		definitionsStatus[idx] = redhatcopv1alpha1.VaultSecretDefinitionStatus{
+			Name:          vaultSecretDefinition.Name,
+			LeaseID:       vaultSecret.LeaseID,
+			LeaseDuration: vaultSecret.LeaseDuration,
+			Renewable:     vaultSecret.Renewable,
+		}
+
+		mergedMap[vaultSecretDefinition.Name] = vaultSecret.Data
+	}
+
+	k8sSecret, err := r.formatK8sSecret(instance, mergedMap)
+	if err != nil {
+		r.Log.Error(err, "unable to format k8s secret", "instance", instance)
+		return err
+	}
+
+	err = r.CreateOrUpdateResource(ctx, instance, instance.GetNamespace(), k8sSecret)
+	if err != nil {
+		return err
+	}
+
+	now := metav1.NewTime(time.Now())
+	instance.Status.LastVaultSecretUpdate = &now
+	instance.Status.VaultSecretDefinitionsStatus = definitionsStatus
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VaultSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	vaultSecretPredicate := predicate.Funcs{
```

Comment:
this is not needed as what you have coded is the default behavior

Reply:
Not sure I understand. I only want update events when the spec changes. I don't want an update if the status field changes, for example. This is not the right way to do it?

Reply:
go it sorry, may bad